File: python/face/api/metodos/interpolacion/lagrange.py
import numpy as np
from sympy import sympify, simplify

from ..numeric_method import NumericMethod
from ..ecuaciones_no_lineales import utils


class Lagrange(NumericMethod):
    def calculate(self, parameters):
        x = parameters["X"]
        y = parameters["Y"]
        x_eval = parameters["eval"]

        x = self.process_params(x)
        y = self.process_params(y)

        respuesta = ""
        n = len(x)
        res = ["" for x in range(n)]
        numerador = ""
        denominador = ""
        polinomio = ""

        # Se crea el arreglo, cada posicion es un y*Li(x)
        for i in range(0, n):
            numerador = ""
            denominador = ""
            for j in range(0, n):
                if j is not i:
                    numerador += "(x" + "-" + str(x[j]) + ")*"
                    denominador += ("("+str(x[i]) + "-" + str(x[j]) + ")*")
            numerador = numerador[:-1]
            denominador = denominador[:-1]
            aux = str(y[i]) + "*(" + numerador + ")/(" + denominador + ")"
            res[i] = aux

        # Se suman todos los y*Li(x)
        for i in res:
            polinomio += "(" + i + ")" + "+"
        polinomio = polinomio[:-1]

        polinomio = sympify(polinomio)
        polinomio = simplify(polinomio)
        polinomio = str(polinomio)
        respuesta = "p(x) = " + polinomio

        y_eval = utils.eval_f(polinomio, x_eval)
        y_eval = round(eval(y_eval), 2)

        return {"funcion": respuesta, "y_eval": y_eval}

    def process_params(self, array):
        n = len(array)
        new_vector = np.zeros(n)

        for i in range(n):
            new_vector[i] = array[str(i)]

        return new_vector

File: java/face/src/main/java/co/edu/eafit/interpolation/LaGrange.java
package co.edu.eafit.interpolation;

import co.edu.eafit.NumericMethod;
import co.edu.eafit.Solution;

public class LaGrange implements NumericMethod {

	@Override
	public Solution calculate(Object... args) {
		double[] x = (double[]) args[0];
		double[] y = (double[]) args[1];
		
		int n = x.length;
		String[] res = new String[n];
		
		String numerador;
		String denominador;
		String aux;
		
		for (int i = 0; i < n; i++) {
			numerador = "";
			denominador = "";
			
			for (int j = 0; j < res.length; j++) {
				if (j != i) {
					numerador += "(x-" + x[j] + ")*";
					denominador += "(" + x[i] + "-" + x[j] +")*";
				}
			}
			
			numerador = numerador.substring(0, numerador.length()-1);
			denominador= denominador.substring(0, denominador.length()-1);
			
			aux = y[i] + "*" + numerador + "/" + denominador;
			res[i] = aux;
		}
		
		aux = "p(x) = ";
		for (String ecuacion : res) {
			aux += "(" + ecuacion + ")" + "+";
		}
		
		aux = aux.substring(0, aux.length()-1);
		
		return new InterpolationSolution(aux, false);
	}

	@Override
	public String getName() {
		return "Polinomio de LaGrange";
	}

}
